Calculator

RomanLevashev opened the pull request.

RomanLevashev committed:
Calculator

RomanLevashev committed:
Culture changed to ru

RomanLevashev committed:
Culture changed to ru in CalculatorLogic

yurii-litvinov commented on Calculator/CalculatorGUI/CalculatorForm.cs:
```diff
@@ -0,0 +1,45 @@
+namespace CalculatorGUI
```
- Используйте file-scoped namespaces
- Нужна ещё шапка с информацией о лицензии

yurii-litvinov commented on Calculator/CalculatorGUI/CalculatorGUI.csproj:
```diff
@@ -0,0 +1,15 @@
+
+  <PropertyGroup>
+    <OutputType>WinExe</OutputType>
+    <TargetFramework>net8.0-windows</TargetFramework>
```
Пора переходить на свежие версии .NET

yurii-litvinov commented on Calculator/CalculatorGUI/CalculatorForm.cs:
```diff
@@ -0,0 +1,45 @@
+                {
+                    Button clickedButton = (Button)sender!;
+                    (string newExpression, string displayResult, bool isChanged) = 
+                    calculator.UpdateExpression((ButtonType)clickedButton.Tag!, clickedButton.Text[0]);
```
UI мог бы взять на себя сортировку кнопок, чтобы не в UpdateExpression был гигантский switch, а тут несколько обработчиков (один для всех цифровых кнопок и по одному для каждой из остальных), и для каждого — свой метод в Calculator. Вам же всё равно надо различать выполняемые действия, пусть это делает событийный механизм в UI, а не руками в бэкенде. И ещё это позволит избавиться от жёсткой привязки бэкенда к строкам-тегам в UI.

yurii-litvinov commented on Calculator/CalculatorLogic/Calculator.cs:
```diff
@@ -0,0 +1,162 @@
+                return;
+            }
+
+            string[] tokens = expression.Split(' ');
```
Мы сначала склеиваем строки, потом тут парсим? Не лучший способ бесцельно потратить усилия :)

yurii-litvinov commented on Calculator/CalculatorLogic/CalculatorLogic.csproj:
```diff
@@ -0,0 +1,16 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118">
```
Да и StyleCop неплохо бы обновить

yurii-litvinov commented on Calculator/CalculatorTests/CalculatorLogicTest.cs:
```diff
@@ -0,0 +1,124 @@
+        [TestMethod]
+        public void ZeroDivisonTest()
+        {
+            Calculator calculator = new();
```
Вынесли бы в SetUp, чтобы не писать каждый раз

yurii-litvinov reviewed (changes requested):
Форма называется не очень, и нет кнопки сброса :) Но в целом круто.

RomanLevashev committed:
Merge branch 'main' into Calculator

RomanLevashev committed:
Added license headers to all files
Refactored input handling: Button type validation moved from CalculatorLogic to CalculatorGUI.
Replaced UpdateExpression with dedicated methods for different operations.
Enhanced expression parser logic in CalculatorLogic.
Updated tests to align with the new architecture.
Upgraded to the latest StyleCop version and .NET 9

yurii-litvinov commented on Calculator/CalculatorGUI/CalculatorForm.cs:
```diff
@@ -0,0 +1,81 @@
+                    case ButtonType.Delete:
+                        isChanged = calculator.Delete();
+                        break;
+                }
```
Идея про "несколько обработчиков (один для всех цифровых кнопок и по одному для каждой из остальных)" из комментариев к прошлой версии была либо не так понята, либо не так реализована :) Я имел в виду что-то вроде `this.delete.Click += calculator.Delete();` и т.п.

yurii-litvinov reviewed (approved):
Окей, зачтена
